test/integration/framework/etcd.go

```diff
@@ -97,36 +90,45 @@ func init() {
 	grpclog.SetLoggerV2(grpclog.NewLoggerV2(io.Discard, io.Discard, os.Stderr))
 }
 
+// hiddenEtcdMessages contains all messages that occur due to the way we run
+// etcd and don't need to be logged in all tests.
+var hiddenEtcdMessages = sets.New(
+	"simple token is not cryptographically signed",
+	"Running http and grpc server on single port. This is not recommended for production.",
+)
+
 // RunCustomEtcd starts a custom etcd instance for test purposes.
-func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url string, stopFn func(), err error) {
+func RunCustomEtcd(logger klog.Logger, dataDir string, customFlags []string) (url string, stopFn func(), err error) {
 	// TODO: Check for valid etcd version.
 	etcdPath, err := getEtcdPath()
 	if err != nil {
 		fmt.Fprint(os.Stderr, installEtcd)
 		return "", nil, fmt.Errorf("could not find etcd in PATH: %v", err)
 	}
-	etcdPort, err := getAvailablePort()
-	if err != nil {
-		return "", nil, fmt.Errorf("could not get a port: %v", err)
-	}
-	customURL := fmt.Sprintf("http://127.0.0.1:%d", etcdPort)
-
-	klog.Infof("starting etcd on %s", customURL)
-
 	etcdDataDir, err := os.MkdirTemp(os.TempDir(), dataDir)
 	if err != nil {
 		return "", nil, fmt.Errorf("unable to make temp etcd data dir %s: %v", dataDir, err)
 	}
-	klog.Infof("storing etcd data in: %v", etcdDataDir)
+	etcdSocketPath := path.Join(etcdDataDir, "etcd.sock")
+	customURL := "unix://" + etcdSocketPath
 
+	logger.V(2).Info("starting etcd", "url", customURL, "dataDir", etcdDataDir)
 	ctx, cancel := context.WithCancel(context.Background())
 	args := []string{
 		"--data-dir",
 		etcdDataDir,
 		"--listen-client-urls",
 		customURL,
+		// This should be how clients connect to etcd, but https://github.com/etcd-io/etcd/pull/12469
+		// apparently was incomplete: trying to pass a Unix Domain URL here is rejected by ectd 3.15.13 with
+		//    --advertise-client-urls "unix:///tmp/etcd.sock" must be "host:port" (missing port in address)
+		//
+		// We don't need to advertise the correct address. To prevent connecting to the default URL
+		// in the unlikely case that something does use this URL after all, an invalid URL is set here.
 		"--advertise-client-urls",
-		customURL,
+		"http://127.0.0.111:0",
```

Comment:
but does 127.0.0.111 exist?

Reply:
It doesn't matter, the whole point is that this can't be the real Unix Domain socket because etcd doesn't allow it and then better should be something which is guaranteed to be invalid (because of the 0 port number). This is not for listening, it's for clients.

Reply:
It has a default value of `http://localhost:2379` which then would be used.

None of this matters unless some client somehow picks up the advertised URL and connects to it. This should never happen as far as I understand it. But if it happens, I'd rather have that client fail than potentially connect to some actual etcd instance.

Not important, I can drop it.

Reply:
Maybe we should pick an address we _know_ won't work but will technically be a valid value. Off the top of my head one of the well known DNS resolver IPs could work...

Reply:
> if we can't unset it to ""

I checked, `--advertise-client-urls=` didn't override the default.

> Maybe we should pick an address we know won't work but will technically be a valid value.

The sooner a client fails which tries to use the advertised URL the better, IMHO. Therefore I prefer an invalid address.

Reply:
I think 127.0.0.111 meets "pretty unlikely" but not "all but guaranteed". Something like 8.8.8.8 might make more sense.

I also thought about an invalid domain but IIRC Etcd doesn't let you advertise a hostname, only IP?

Reply:
I misspoke: I meant "invalid URL", not "invalid address". The `:0` port is unusable for a connect call.